docs/tutorials/fault_tolerance.md

```diff
@@ -0,0 +1,154 @@
+you can have with RocksDB is limited only by the size of the local disk. Note also that only the
+RocksDBStateBackend is able to do incremental snapshotting, which is a significant benefit for
+applications with large amounts of slowly changing state.
+
+Both state backends are able to do asynchronous snapshotting, meaning that they can take a snapshot without impeding the ongoing stream processing.
+
+{% top %}
+
+## State Snapshots
+
+### Definitions
+
+* _Snapshot_ -- a generic term referring to a global, consistent image of the state of a Flink job.
+  A snapshot includes a pointer into each of the data sources (e.g., an offset into a file or Kafka
+  partition), as well as a copy of the state from each of the job's stateful operators that resulted
+  from having processed all of the events up to those positions in the sources.
+* _Checkpoint_ -- a snapshot taken automatically by Flink for the purpose of being able to recover
+  from faults. Checkpoints can be incremental, and are optimized for being restored quickly.
+* _Externalized Checkpoint_ -- normally checkpoints are not intended to be manipulated by users.
+  Flink retains only the _n_-most-recent checkpoints (_n_ being configurable) while a job is
+  running, and deletes them when a job is cancelled. But you can configure them to be retained
+  instead, in which case you can manually resume from them.
+* _Savepoint_ -- a snapshot triggered manually by a user (or an API call) for some operational
+  purpose, such as a stateful redeploy/upgrade/rescaling operation. Savepoints are always complete,
+  and are optimized for operational flexibility.
+
+### How does State Snapshotting Work?
+
+Flink uses a variant of the [Chandy-Lamport
+algorithm](https://en.wikipedia.org/wiki/Chandy-Lamport_algorithm) known as _asynchronous barrier
+snapshotting_.
+
+When a task manager is instructed by the checkpoint coordinator (part of the job manager) to begin a
+checkpoint, it has all of the sources record their offsets and insert numbered _checkpoint barriers_
+into their streams. These barriers flow through the job graph, indicating the part of the stream
+before and after each checkpoint. 
+
+<img src="{{ site.baseurl }}/fig/stream_barriers.svg" alt="Checkpoint barriers are inserted into the streams" class="center" width="80%" />
+
+Checkpoint _n_ will contain the state of each operator that resulted from having consumed **every
+event before checkpoint barrier _n_, and none of the events after it**.
+
+As each operator in the job graph receives one of these barriers, it records its state. Operators
+with two input streams (such as a CoProcessFunction) perform _barrier alignment_ so that the
+snapshot will reflect the state resulting from consuming events from both input streams up to (but
+not past) both barriers.
+
+<img src="{{ site.baseurl }}/fig/stream_aligning.svg" alt="Barrier alignment" class="center" width="80%" />
```

Comment:
I think, this should be at least 90%, better 100% for better readability; the fonts are quite small

Reply:
agreed; done